sdk/eventhub/Azure.Messaging.EventHubs.Processor/src/EventProcessorClient.cs

```diff
@@ -1092,48 +1040,138 @@ protected override async Task<EventProcessorCheckpoint> GetCheckpointAsync(strin
         }
 
         /// <summary>
-        ///   Creates a <see cref="StorageManager" /> to use for interacting with durable storage.
-        /// </summary>
-        ///
-        /// <param name="checkpointStore">The client responsible for interaction with durable storage, responsible for persisting checkpoints and load-balancing state.</param>
-        ///
-        /// <returns>A <see cref="StorageManager" /> with the requested configuration.</returns>
-        ///
-        private StorageManager CreateStorageManager(BlobContainerClient checkpointStore) => new BlobsCheckpointStore(checkpointStore, RetryPolicy);
-
-        /// <summary>
-        ///   Processes the starting positions for checkpoints, ensuring that any overrides set by the <see cref="PartitionInitializingAsync" />
-        ///   handler are respected when no natural checkpoint exists for the partition.
+        ///   Signals the <see cref="EventProcessorClient" /> to begin processing events.  Should this method be called while the processor
+        ///   is running, no action is taken.
         /// </summary>
         ///
-        /// <param name="sourceCheckpoints">The checkpoint set to process.</param>
+        /// <param name="async">When <c>true</c>, the method will be executed asynchronously; otherwise, it will execute synchronously.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the start operation.  This won't affect the <see cref="EventProcessor{TPartition}" /> once it starts running.</param>
         ///
-        /// <returns>An enumerable consisting of the <paramref name="sourceCheckpoints" /> and a set of artificial checkpoints for any overrides applied to the starting position.</returns>
-        ///
-        private IEnumerable<EventProcessorCheckpoint> ProcessCheckpointStartingPositions(IEnumerable<EventProcessorCheckpoint> sourceCheckpoints)
+        private async Task StartProcessingInternalAsync(bool async,
+                                                        CancellationToken cancellationToken)
         {
-            var knownCheckpoints = new HashSet<string>();
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+
+            var capturedValidationException = default(Exception);
+            var releaseGuard = false;
+
+            try
+            {
+                // Acquire the semaphore used to synchronize processor starts and stops, respecting
+                // the async flag.  When this is held, the state of the processor is stable.
+
+                if (async)
+                {
+                    await ProcessorStatusGuard.WaitAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
Is there a reasonable timeout that could be used here just in case this waits forever?

Reply:
I'm going to leave this as-is for now; the cancellation token gives callers control over timeouts and I'm wary of changing the existing startup behavior in a way that may impact applications already using the processor in an unexpected way.   The best we could do here would be abort and throw.    I'll think on this a bit more over the next couple of days.